### PR TITLE
Move conversion of QuasiDistribution to result decoder

### PR DESCRIPTION
**Pull request by jyu00**
### Summary

#455 converted `SamplerResult.quasi_dists` from dict to `QuasiDistribution` to match the `SamplerResult` description. That PR was written before we deprecated `_call` and moved to `_run`, so it only applied to the old path. It is also why the latest integration test failed.

This PR moved that code to `SamplerResultDecoder`, which does result processing when `job.result()` is called. 

@t-imamichi Are you ok if I use the old code in the old path - i.e. the result returned by `_call()` will be `SamplerResult` with a list of dict instead of `QuasiDistribution`? This way it's no longer a breaking change, and people using the new `run()` method will get the correct output (list of `QuasiDistribution`). 

### Details and comments
Fixes #

### CODE REVIEW

**Review by kt474 — APPROVED** [review]
LGTM, can wait for @t-imamichi's approval

**Comment by t-imamichi** [comment]
Thanks. LGTM. We need to simplify the code when we remove the old path though.

**Pull request merged**
